### PR TITLE
add mxfp8 padding and tests for esm-2

### PR DESCRIPTION
**Pull request by pstjohn**
Adds padding to the vocab dimension of ESM-2 to make it compatible with FP8 and MXFP8 kernels

## Summary by CodeRabbit

- **New Features**
  - FP8/MXFP8 execution support with parity checks; configurable padded vocabulary (padded_vocab_size, default 64) enabling FP8-friendly embeddings/LM outputs while truncating outputs back to the original vocab size.
- **Refactor**
  - Simplified, PyTorch-centric data-collation pipeline; collators enforce PyTorch tensors and support padding to a specified multiple.
- **Documentation**
  - Config docs updated to describe padded_vocab_size and its constraints.
- **Tests**
  - Added FP8/MXFP8 forward-pass tests, collator/unit tests, and state-dict padding validation; test fixtures reorganized for a shared converted checkpoint.

### CODE REVIEW

File: bionemo-recipes/models/esm2/src/esm/collator.py [hunk 1]
```diff
@@ -130,205 +153,197 @@ def __call__(self, features, return_tensors=None):
...
+            [
+                batch["cu_seq_lens_q"],
+                torch.tensor([batch["cu_seq_lens_q"][-1] + remainder], dtype=batch["cu_seq_lens_q"].dtype),
+            ],
+            dim=0,
+        )
+        batch["cu_seq_lens_k"] = batch["cu_seq_lens_q"]
+
+    if "max_length_q" in batch:
+        batch["max_length_q"] = max(batch["max_length_q"], remainder)
```
Comment (jomitchellnv) [inline bionemo-recipes/models/esm2/src/esm/collator.py:341]: So the `remainder` is a separate entry in the input_ids right? So that's why we check to see if the len(remainder) is > len(current_max_length) here right?

File: bionemo-recipes/models/esm2/src/esm/collator.py [hunk 2]
```diff
@@ -130,205 +153,197 @@ def __call__(self, features, return_tensors=None):
...
+        batch["attention_mask"] = torch.tensor(
+            [[v for sample in features for v in sample["attention_mask"]]], dtype=torch.int64
+        )
+    return batch
+
+
+def _pt_pad_to_multiple_of(batch: dict[str, Any], pad_to_multiple_of: int, token_pad: int, label_pad: int):
+    """Pad a batch to a multiple of pad_to_multiple_of.
+
+    Appends a mock sequence to the end of the batch with the given token_pad and label_pad to make the total number of
```
Comment (jomitchellnv) [inline bionemo-recipes/models/esm2/src/esm/collator.py:303]: It adds multiple mock sequences right? One to the `input_ids`, another to the `labels`. 

Also, what about the positional_ids? We need to add them as well, since now we have no positional_ids for the new sequence right?

File: bionemo-recipes/models/esm2/src/esm/collator.py [hunk 3]
```diff
@@ -130,205 +153,197 @@ def __call__(self, features, return_tensors=None):
...
+    cu_seq_lens[1:] = torch.cumsum(torch.tensor(sample_lengths), dim=0, dtype=torch.int32)
+    batch["cu_seq_lens_q"] = batch["cu_seq_lens_k"] = cu_seq_lens
+    if "attention_mask" in features[0]:
+        batch["attention_mask"] = torch.tensor(
+            [[v for sample in features for v in sample["attention_mask"]]], dtype=torch.int64
+        )
+    return batch
+
+
+def _pt_pad_to_multiple_of(batch: dict[str, Any], pad_to_multiple_of: int, token_pad: int, label_pad: int):
```
Comment (jomitchellnv) [inline bionemo-recipes/models/esm2/src/esm/collator.py:300]: It's probably cleaner if we pass the actual arguments such as `input_ids`, `labels`, etc instead of the `batch` dictionary. I can move this into TransformerEngine utilities later if we want also.

File: bionemo-recipes/models/esm2/src/esm/convert.py [hunk 4]
```diff
@@ -114,3 +111,40 @@ def _pack_qkv_bias(ctx: io.TransformCTX, query, key, value):
     concat_biases = concat_biases.transpose(0, 1).contiguous()
     concat_biases = concat_biases.view(*input_shape)
     return concat_biases
+
+
+def _pad_weights(ctx: io.TransformCTX, source_embed):
```
Comment (jomitchellnv) [inline bionemo-recipes/models/esm2/src/esm/convert.py:116]: missing docstring about what source_embed is. It looks like we're padding the vocab size here, i'm not sure why though. Isn't the vocab size just for looking up values but doesn't directly affect the intermediate shape?

File: bionemo-recipes/models/esm2/src/esm/convert.py [hunk 5]
```diff
@@ -114,3 +111,40 @@ def _pack_qkv_bias(ctx: io.TransformCTX, query, key, value):
...
+    source_key="lm_head.decoder.weight",
+    target_key="lm_head.decoder.weight",
+)(_pad_weights)
+
+
+@io.state_transform(
+    source_key="lm_head.bias",
+    target_key="lm_head.decoder.bias",
+)
+def _pad_bias(ctx: io.TransformCTX, source_bias):
```
Comment (jomitchellnv) [inline bionemo-recipes/models/esm2/src/esm/convert.py:142]: not sure what's going on here or why this is necessary

Comment (pstjohn) [reply bionemo-recipes/models/esm2/src/esm/convert.py:142]: that final linear layer dimension needs to be divisible by 64 for mxfp8; but the original bias is only 33. So we need to pad it out with zeros

File: bionemo-recipes/models/esm2/src/esm/collator.py (hunk 2 above)
Comment (pstjohn) [reply bionemo-recipes/models/esm2/src/esm/collator.py:303]: we add a single mock sequence (e.g., protein sequence), that has both input_ids and labels. 
we don't currently return positional_ids. Do we need them?

File: bionemo-recipes/models/esm2/src/esm/collator.py (hunk 1 above)
Comment (pstjohn) [reply bionemo-recipes/models/esm2/src/esm/collator.py:341]: "remainder" is the length of the pseudo-sequence we're adding

File: bionemo-recipes/models/esm2/src/esm/convert.py (hunk 4 above)
Comment (pstjohn) [reply bionemo-recipes/models/esm2/src/esm/convert.py:116]: it's tied to the decoder weights
